src/stim/util_top/reference_sample_tree.cc

```diff
@@ -0,0 +1,179 @@
+            .repetitions = 1,
+        });
+    }
+    for (const auto &child : suffix_children) {
+        child.flatten_and_simplify_into(flattened);
+    }
+
+    // Fuse children.
+    auto &f = flattened;
+    for (size_t k = 0; k < f.size(); k++) {
+        // Combine children with identical contents by adding their rep counts.
+        while (k + 1 < f.size() && f[k].prefix_bits == f[k + 1].prefix_bits && f[k].suffix_children == f[k + 1].suffix_children) {
+            f[k].repetitions += f[k + 1].repetitions;
+            f.erase(f.begin() + k + 1);
+        }
+
+        // Fuse children with unrepeated contents if they can be fused.
+        while (k + 1 < f.size() && f[k].repetitions == 1 && f[k].suffix_children.empty() && f[k + 1].repetitions == 1) {
+            f[k].suffix_children = std::move(f[k + 1].suffix_children);
+            f[k].prefix_bits.insert(f[k].prefix_bits.end(), f[k + 1].prefix_bits.begin(), f[k + 1].prefix_bits.end());
+            f.erase(f.begin() + k + 1);
+        }
+    }
+
+    if (repetitions == 1) {
+        // Un-nest all the children.
+        for (auto &e : flattened) {
+            out.push_back(e);
+        }
+    } else if (flattened.size() == 1) {
+        // Merge with single child.
+        flattened[0].repetitions *= repetitions;
+        out.push_back(std::move(flattened[0]));
+    } else if (flattened.empty()) {
+        // Nothing to report.
+    } else if (flattened[0].suffix_children.empty() && flattened[0].repetitions == 1) {
+        // Take payload from first child.
+        auto result = std::move(flattened[0]);
+        flattened.erase(flattened.begin());
+        result.repetitions = repetitions;
+        result.suffix_children = std::move(flattened);
+        out.push_back(std::move(result));
+    } else {
+        out.push_back(ReferenceSampleTree{
+            .prefix_bits={},
+            .suffix_children=std::move(flattened),
+            .repetitions=repetitions,
+        });
+    }
+}
+
+void ReferenceSampleTree::try_factorize(size_t period_factor) {
+    if (prefix_bits.size() != 0 || suffix_children.size() % period_factor != 0) {
+        return;
+    }
+
+    // Check if contents are periodic with the factor.
+    size_t h = suffix_children.size() / period_factor;
+    for (size_t k = h; k < suffix_children.size(); k++) {
+        if (suffix_children[k - h] != suffix_children[k]) {
+            return;
+        }
+    }
+
+    // Factorize.
+    suffix_children.resize(h);
+    repetitions *= period_factor;
+}
+
+ReferenceSampleTree ReferenceSampleTree::simplified() const {
+    std::vector<ReferenceSampleTree> flat;
+    flatten_and_simplify_into(flat);
+    if (flat.empty()) {
+        return ReferenceSampleTree();
+    } else if (flat.size() == 1) {
+        return flat[0];
+    }
+
+    ReferenceSampleTree result;
+    result.repetitions = 1;
+
+     // Take payload from first child.
+    if (flat[0].repetitions == 1 && flat[0].suffix_children.empty()) {
```

Comment:
Unfortunately although the logic is very similar they actually apply to very slightly different cases. IIRC the tests fail if you delete either one. Definitely fail if you delete this one.

Reply:
This might be due to the recursive aspect that I'm mentioning above. It just so happens that your test cases require this specific unwrapping step to be applied twice, but any of them could be necessary any number of times.

Reply:
Ok, after looking at this more. I see that this could never be fully covered by the checks in `flatten_and_simplify_into()` simply due to the differences is return type. This has to wrap the vector in a tree, which could then in it of itself be simplified. I think its clear that there's nothing else to change without changing the return type of these methods (to be the same thing) or something similar.